[UnifiedPDF][iOS] Start zooming in the plugin.

sammygill opened the pull request:
#### bc3a3d94e49ea4c9bbef02dba14d71be8191c540
<pre>
[UnifiedPDF][iOS] Start zooming in the plugin.
<a href="https://bugs.webkit.org/show_bug.cgi?id=279689">https://bugs.webkit.org/show_bug.cgi?id=279689</a>
<a href="https://rdar.apple.com/120425065">rdar://120425065</a>

Reviewed by Abrar Rahman Protyasha and Wenson Hsieh.

This patch serves as a first step towards allowing the plugin to be completely responsible
for zooming on iOS and focuses on allowing this behavior for main frame PDFs.

When the UIScrollView is zoomed and the delegate is informed of the zoom events we will
check to see if the web content process has a main framed plugin that is responsible for
handling the page scale gesture. If it does we will inform it about the beginning of
the gesture and the end. This information gets plumbed to the PDFPlugin which keeps track
of it as an extra bit of state.

Whenever the WebPage is handling a content rect update from the UIProcess we will forward
over the new page scale to the plugin to handle rather than sending it over to the page.
The origin that is passed along to the plugin is just the origin of the visible content
rect that we got from the UIProcess so we will just simply scroll to its position in
the ScrolledContents coordinate space.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView scrollViewWillBeginZooming:withView:]):
(-[WKWebView scrollViewDidZoom:]):
(-[WKWebView scrollViewDidEndZooming:withView:atScale:]):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::didEndUserTriggeredZooming):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::setScaleFactor):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::updateVisibleContentRects):
(WebKit::WebPage::willStartUserTriggeredZooming):
(WebKit::WebPage::didEndUserTriggeredZooming):

Canonical link: <a href="https://commits.webkit.org/283888@main">https://commits.webkit.org/283888@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b585e738096c6b2078d72d398138e984781ae531

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/67680 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/47062 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/20319 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/71735 "Built successfully") | [⏳ 🛠 wincairo ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/54861 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/18624 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/54170 "Passed tests") | [⏳ 🧪 wincairo-tests ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/70747 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/43163 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/58525 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/34638 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39837 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/17179 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/61801 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/16272 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/73433 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/11643 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/15579 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/61617 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/11678 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/58593 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/61666 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15036 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/9499 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/3114 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/42865 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/43943 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/45132 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/43684 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -4620,7 +4621,11 @@ static bool selectionIsInsideFixedPositionContainer(LocalFrame& frame)
 
-            m_page->setPageScaleFactor(scaleFromUIProcess.value(), scrollPosition, m_isInStableState);
+            if (auto* pluginView = mainFramePlugIn())
+                pluginView->setPageScaleFactor(scaleFromUIProcess.value(), std::nullopt);
```
I think the scrollPosition is the origin we're passing to `WebCore::Page::setPageScaleFactor()`, so shouldn't we do likewise when invoking `PluginView::setPageScaleFactor()`?

sammygill replied:
Yes indeed! Changed!

aprotyas commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -4620,7 +4621,11 @@ static bool selectionIsInsideFixedPositionContainer(LocalFrame& frame)
             m_dynamicSizeUpdateHistory.clear();
 
-            m_page->setPageScaleFactor(scaleFromUIProcess.value(), scrollPosition, m_isInStableState);
+            if (auto* pluginView = mainFramePlugIn())
```
Also, `#if ENABLE(PDF_PLUGIN)`!

```cpp
#if ENABLE(PDF_PLUGIN)
if (auto* pluginView = mainFramePlugin())
    pluginView->setPageScaleFactor(...);
else
#endif
    m_page->setPageScaleFactor(...);
```

whsieh commented on Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
```diff
@@ -552,6 +552,7 @@ GenerateSyntheticEditingCommand(enum:uint8_t WebKit::SyntheticEditingCommandType
 
 #if PLATFORM(IOS_FAMILY)
     WillStartUserTriggeredZooming();
+    didEndUserTriggeredZooming();
```
```suggestion
    DidEndUserTriggeredZooming();
```

whsieh commented on Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
```diff
@@ -779,6 +779,11 @@ static inline float adjustedUnexposedMaxEdge(float documentEdge, float exposedRe
 
+void WebPageProxy::didEndUserTriggeredZooming()
+{
+    legacyMainFrameProcess().send(Messages::WebPage::didEndUserTriggeredZooming(), webPageIDInMainFrameProcess());
```
```suggestion
    legacyMainFrameProcess().send(Messages::WebPage::DidEndUserTriggeredZooming(), webPageIDInMainFrameProcess());
```

whsieh commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -4620,7 +4621,13 @@ static bool selectionIsInsideFixedPositionContainer(LocalFrame& frame)
 
-            m_page->setPageScaleFactor(scaleFromUIProcess.value(), scrollPosition, m_isInStableState);
+#if ENABLE(PDF_PLUGIN)
+            if (auto* pluginView = mainFramePlugIn())
```
Smart pointer nit -

```suggestion
            if (RefPtr pluginView = mainFramePlugIn())
```

whsieh commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -4788,10 +4795,21 @@ static bool selectionIsInsideFixedPositionContainer(LocalFrame& frame)
 
 void WebPage::willStartUserTriggeredZooming()
 {
+    if (auto* pluginView = mainFramePlugIn()) {
```
```suggestion
    if (RefPtr pluginView = mainFramePlugIn()) {
```

whsieh commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -4788,10 +4795,21 @@ static bool selectionIsInsideFixedPositionContainer(LocalFrame& frame)
 
+void WebPage::didEndUserTriggeredZooming()
+{
+    if (auto* pluginView = mainFramePlugIn())
```
```suggestion
    if (RefPtr pluginView = mainFramePlugIn())
```

aprotyas reviewed (approved):
r+ with Wenson's comments. Also, I think we need a few more `ENABLE(PDF_PLUGIN)` guards in `WebPageIOS.mm`.

sammygill committed:
[UnifiedPDF][iOS] Start zooming in the plugin.
https://bugs.webkit.org/show_bug.cgi?id=279689
rdar://120425065

Reviewed by Abrar Rahman Protyasha and Wenson Hsieh.

This patch serves as a first step towards allowing the plugin to be completely responsible
for zooming on iOS and focuses on allowing this behavior for main frame PDFs.

When the UIScrollView is zoomed and the delegate is informed of the zoom events we will
check to see if the web content process has a main framed plugin that is responsible for
handling the page scale gesture. If it does we will inform it about the beginning of
the gesture and the end. This information gets plumbed to the PDFPlugin which keeps track
of it as an extra bit of state.

Whenever the WebPage is handling a content rect update from the UIProcess we will forward
over the new page scale to the plugin to handle rather than sending it over to the page.
The origin that is passed along to the plugin is just the origin of the visible content
rect that we got from the UIProcess so we will just simply scroll to its position in
the ScrolledContents coordinate space.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView scrollViewWillBeginZooming:withView:]):
(-[WKWebView scrollViewDidZoom:]):
(-[WKWebView scrollViewDidEndZooming:withView:atScale:]):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::didEndUserTriggeredZooming):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::setScaleFactor):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::updateVisibleContentRects):
(WebKit::WebPage::willStartUserTriggeredZooming):
(WebKit::WebPage::didEndUserTriggeredZooming):

Canonical link: https://commits.webkit.org/283888@main
